File: app/src/main/java/com/uncmorfi/balance/UserCursorAdapter.java
package com.uncmorfi.balance;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.RoundedBitmapDrawable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory;
import android.support.v4.view.animation.LinearOutSlowInInterpolator;
import android.support.v7.widget.RecyclerView;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.target.BitmapImageViewTarget;
import com.uncmorfi.R;
import com.uncmorfi.balance.model.User;
import com.uncmorfi.balance.model.UsersContract;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * Llena un RecyclerView.
 * El fragmento o actividad contenedora debería implementar {@link OnCardClickListener}.
 */
class UserCursorAdapter extends RecyclerView.Adapter<UserCursorAdapter.UserItemViewHolder> {
    public static final String USER_IMAGES_URL = "https://asiruws.unc.edu.ar/foto/";
    private static final float SCALE_USER_IMAGE_SIZE = 0.8f;
    private static final int SCALE_USER_IMAGE_TIME = 500;
    private static final int WARNING_USER_BALANCE = 20;
    private static final int WARNING_USER_EXPIRE = 1;
    private DateFormat mDateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.ROOT);
    private Context mContext;
    private Cursor mCursor;
    private List<Boolean> mUpdateInProgress = new ArrayList<>();
    private OnCardClickListener mListener;

    interface OnCardClickListener {
        void onClick(int userId, String userCard, int position);
    }

    class UserItemViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView nameText;
        TextView cardText;
        TextView typeText;
        TextView balanceText;
        ImageView userImage;
        ProgressBar progressBar;
        TextView expirationText;
        TextView lastUpdateText;

        UserItemViewHolder(View v) {
            super(v);

            nameText = v.findViewById(R.id.user_name);
            cardText = v.findViewById(R.id.user_card);
            typeText = v.findViewById(R.id.user_type);
            balanceText = v.findViewById(R.id.user_balance);
            userImage = v.findViewById(R.id.user_image);
            progressBar = v.findViewById(R.id.user_bar);
            expirationText = v.findViewById(R.id.user_expiration);
            lastUpdateText = v.findViewById(R.id.user_last_update);

            v.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int pos = this.getAdapterPosition();
            mListener.onClick(getItemIdFromCursor(pos), getItemCardFromCursor(pos), pos);
        }
    }

    UserCursorAdapter(Context context, OnCardClickListener listener) {
        mContext = context;
        mListener = listener;
    }

    @Override
    public @NonNull UserItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.user_item, parent, false);
        return new UserItemViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull final UserItemViewHolder holder, int position) {
        mCursor.moveToPosition(position);
        User user = new User(mCursor);

        setHolder(holder, user);
        setImage(holder, user);
        setProgressBar(holder, position);
    }

    private void setHolder(UserItemViewHolder holder, User user) {
        holder.nameText.setText(user.getName());
        holder.cardText.setText(user.getCard());
        holder.typeText.setText(user.getType());
        holder.balanceText.setText(String.format(Locale.US, "$ %d", user.getBalance()));

        holder.balanceText.setTextColor(ContextCompat.getColor(mContext,
                user.getBalance() < WARNING_USER_BALANCE ? R.color.accent : R.color.primary_dark));

        holder.expirationText.setText(String.format(
                mContext.getString(R.string.balance_expiration),
                textExpiration(user.getExpiration())));

        holder.expirationText.setTextColor(ContextCompat.getColor(mContext,
                warningExpiration(user.getExpiration()) ? R.color.accent : R.color.secondary_text));

        holder.lastUpdateText.setText(String.format(
                mContext.getString(R.string.balance_last_update),
                DateUtils.getRelativeTimeSpanString(user.getLastUpdate()).toString().toLowerCase()));
    }

    private String textExpiration(long expiration) {
        if (expiration == 0)
            return "?";
        else
            return mDateFormat.format(new Date(expiration));
    }

    private boolean warningExpiration(long expiration) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date());
        cal.add(Calendar.MONTH, WARNING_USER_EXPIRE);
        return (new Date(expiration)).before(cal.getTime());
    }

    private void setImage(final UserItemViewHolder holder, User user) {
        Glide.with(mContext)
                .load(USER_IMAGES_URL + user.getImage())
                .asBitmap()
                .placeholder(R.drawable.person_placeholder)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .centerCrop()
                .into(new BitmapImageViewTarget(holder.userImage) {
                    @Override
                    protected void setResource(Bitmap resource) {
                        RoundedBitmapDrawable circularBitmapDrawable =
                                RoundedBitmapDrawableFactory.create(mContext.getResources(), resource);
                        circularBitmapDrawable.setCircular(true);
                        holder.userImage.setImageDrawable(circularBitmapDrawable);
                    }
                });
    }

    private void setProgressBar(UserItemViewHolder holder, int position) {
        if (mUpdateInProgress.get(position)) {
            holder.progressBar.setVisibility(View.VISIBLE);
            setUserImageUpdatingMode(holder);
        } else {
            holder.progressBar.setVisibility(View.INVISIBLE);
            setUserImageNormalMode(holder);
        }
    }

    private void setUserImageUpdatingMode(UserItemViewHolder holder) {
        holder.userImage.setScaleX(1);
        holder.userImage.setScaleY(1);
        holder.userImage.animate()
                .scaleX(SCALE_USER_IMAGE_SIZE)
                .scaleY(SCALE_USER_IMAGE_SIZE)
                .setInterpolator(new LinearOutSlowInInterpolator())
                .setDuration(SCALE_USER_IMAGE_TIME)
                .start();
    }

    private void setUserImageNormalMode(UserItemViewHolder holder) {
        holder.userImage.setScaleX(SCALE_USER_IMAGE_SIZE);
        holder.userImage.setScaleY(SCALE_USER_IMAGE_SIZE);
        holder.userImage.animate()
                .scaleX(1)
                .scaleY(1)
                .setInterpolator(new LinearOutSlowInInterpolator())
                .setDuration(SCALE_USER_IMAGE_TIME)
                .start();
    }

    @Override
    public int getItemCount() {
        if (mCursor != null)
            return mCursor.getCount();
        return 0;
    }

    void setCursor(Cursor newCursor) {
        mCursor = newCursor;
        resizeUpdateInProgress();
    }

    private void resizeUpdateInProgress() {
        if (mCursor != null) {
            int diff = mCursor.getCount() - mUpdateInProgress.size();
            if (diff > 0) {
                List<Boolean> list = Arrays.asList(new Boolean[diff]);
                Collections.fill(list, Boolean.FALSE);
                mUpdateInProgress.addAll(list);
            }
        }
    }

    void setInProgress(int position, boolean show) {
        mUpdateInProgress.set(position, show);
    }

    private int getItemIdFromCursor(int position) {
        mCursor.moveToPosition(position);
        return mCursor.getInt(mCursor.getColumnIndex(UsersContract.UserEntry._ID));
    }

    String getItemCardFromCursor(int position) {
        mCursor.moveToPosition(position);
        return mCursor.getString(mCursor.getColumnIndex(UsersContract.UserEntry.CARD));
    }
}

File: app/src/main/java/com/uncmorfi/balance/UserCursorAdapter.kt
package com.uncmorfi.balance

import android.content.Context
import android.database.Cursor
import android.graphics.Typeface
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.uncmorfi.R
import com.uncmorfi.balance.model.User
import kotlinx.android.synthetic.main.user_item.view.*
import java.text.SimpleDateFormat
import java.util.*
import android.view.animation.Animation
import android.view.animation.ScaleAnimation
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.request.RequestOptions
import com.uncmorfi.helpers.colorOf

/**
 * Llena un RecyclerView.
 */
internal class UserCursorAdapter(private val mContext: Context,
                                 private val mClickListener: (User) -> Unit,
                                 private val mLongClickListener: (User) -> Unit) :
                                 RecyclerView.Adapter<UserCursorAdapter.UserItemViewHolder>() {
    private val mDateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.ROOT)
    private var mCursor: Cursor? = null
    private val mUpdateInProgress = ArrayList<Boolean>()

    internal inner class UserItemViewHolder(v: View) : RecyclerView.ViewHolder(v) {

        fun bind(user: User) {
            setText(user)
            setImage(user)
            setColors(user)
            setProgressBar(user.position!!)
            itemView.setOnClickListener { mClickListener(user) }
            itemView.setOnLongClickListener { mLongClickListener(user); true }
        }

        private fun setText(user: User) {
            itemView.userName.text = user.name
            itemView.userCard.text = user.card
            itemView.userType.text = user.type

            itemView.userBalance.text = String.format(Locale.US, "$ %d", user.balance)
            itemView.userExpiration.text = mContext.getString(R.string.balance_expiration)
                    .format(textExpiration(user.expiration))
            itemView.userLastUpdate.text = mContext.getString(R.string.balance_last_update)
                    .format(relativeLastUpdate(user.lastUpdate))
        }

        private fun textExpiration(expiration: Long): String {
            return if (expiration == 0L)
                "?"
            else
                mDateFormat.format(Date(expiration))
        }

        private fun relativeLastUpdate(lastUpdate: Long): String {
            return DateUtils.getRelativeTimeSpanString(lastUpdate).toString().toLowerCase()
        }

        private fun setImage(user: User) {
            Glide.with(mContext)
                    .load(user.image)
                    .placeholder(R.drawable.person_placeholder)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .apply(RequestOptions.circleCropTransform())
                    .into(itemView.userImage)
        }

        private fun setColors(user: User) {
            // Alerta si le queda poco saldo
            if (user.balance < WARNING_USER_BALANCE)
                itemView.userBalance.setTextColor(mContext.colorOf(R.color.accent))
            else
                itemView.userBalance.setTextColor(mContext.colorOf(R.color.primary_dark))

            // Alerta si la tarjeta está vencida o está por vencerse
            when {
                expired(user.expiration) -> {
                    itemView.setBackgroundColor(mContext.colorOf(R.color.accent))
                    itemView.userBalance.setTextColor(mContext.colorOf(R.color.white))
                    itemView.userExpiration.setTypeface(null, Typeface.BOLD)
                    itemView.userExpiration.setTextColor(mContext.colorOf(R.color.white))
                    setTextColor(R.color.white, R.color.white)
                }
                warningExpiration(user.expiration) -> {
                    itemView.setBackgroundColor(mContext.colorOf(R.color.white))
                    itemView.userExpiration.setTypeface(null, Typeface.NORMAL)
                    itemView.userExpiration.setTextColor(mContext.colorOf(R.color.accent))
                    setTextColor(R.color.primary_text, R.color.secondary_text)
                }
                else -> {
                    itemView.setBackgroundColor(mContext.colorOf(R.color.white))
                    itemView.userExpiration.setTypeface(null, Typeface.NORMAL)
                    itemView.userExpiration.setTextColor(mContext.colorOf(R.color.secondary_text))
                    setTextColor(R.color.primary_text, R.color.secondary_text)
                }
            }
        }

        private fun setTextColor(primary: Int, extra: Int) {
            itemView.userName.setTextColor(mContext.colorOf(primary))
            itemView.userCard.setTextColor(mContext.colorOf(primary))

            itemView.userType.setTextColor(mContext.colorOf(extra))
            itemView.userLastUpdate.setTextColor(mContext.colorOf(extra))
        }

        private fun warningExpiration(expiration: Long): Boolean {
            val cal = Calendar.getInstance()
            cal.time = Date()
            cal.add(Calendar.MONTH, WARNING_USER_EXPIRE)
            return Date(expiration).before(cal.time)
        }

        private fun expired(expiration: Long): Boolean {
            return Date(expiration).before(Date())
        }

        private fun setProgressBar(position: Int) {
            if (mUpdateInProgress[position]) {
                itemView.userBar.visibility = View.VISIBLE
                scaleView(itemView.userImage, SCALE_USER_IMAGE_SIZE, SCALE_USER_IMAGE_SIZE)
            } else {
                itemView.userBar.visibility = View.INVISIBLE
                scaleView(itemView.userImage,1f,1f)
            }
        }

        private fun scaleView(v: View, endX: Float, endY: Float) {
            val anim = ScaleAnimation(
                    v.scaleX, endX,
                    v.scaleY, endY,
                    Animation.RELATIVE_TO_SELF, 0.5f,
                    Animation.RELATIVE_TO_SELF, 0.5f)
            anim.fillAfter = true
            anim.duration = SCALE_USER_IMAGE_TIME
            v.startAnimation(anim)
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserItemViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.user_item, parent, false)
        return UserItemViewHolder(v)
    }

    override fun onBindViewHolder(holder: UserItemViewHolder, position: Int) {
        mCursor?.let {
            it.moveToPosition(position)
            val user = User(it)
            user.position = position
            holder.bind(user)
        }
    }

    override fun getItemCount(): Int {
        return mCursor?.count ?: 0
    }

    fun setCursor(newCursor: Cursor?) {
        mCursor = newCursor
        resizeUpdateInProgress()
    }

    private fun resizeUpdateInProgress() {
        val cursor = mCursor
        if (cursor != null) {
            val diff = cursor.count - mUpdateInProgress.size
            if (diff > 0) {
                val list = Arrays.asList<Boolean>(*arrayOfNulls(diff))
                list.fill(false)
                mUpdateInProgress.addAll(list)
            }
        }
    }

    fun setInProgress(position: Int, show: Boolean) {
        mUpdateInProgress[position] = show
    }

    companion object {
        private const val SCALE_USER_IMAGE_SIZE = 0.8f
        private const val SCALE_USER_IMAGE_TIME = 500L
        private const val WARNING_USER_BALANCE = 20
        private const val WARNING_USER_EXPIRE = 1
    }
}
